File: python/binary-search.py
import unittest


def binary_search(arr, target):
    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = (left + right) // 2

        if arr[mid] == target:
            return mid
        if arr[mid] < target:
            left = mid + 1
        if arr[mid] > target:
            right = mid

    return -1


sample = [3, 7, 9, 11, 12, 15, 16, 18, 20]

index = binary_search(sample, 7)

print(index)


class SearchTest(unittest.TestCase):
    def test_binary_search(self):
        sample_array = [3, 7, 9, 11, 12, 15, 16, 18, 20]
        target = 7
        self.assertEqual(binary_search(sample_array, target), 1)

    def test_binary_search_will_return_negative_when_number_not_found(self):
        sample_array = [3, 7, 9, 11, 12, 15, 16, 18, 20]
        target = 8
        self.assertEqual(binary_search(sample_array, target), -1)


if __name__ == "__main__":
    unittest.main()

File: golang/binary-search.go
package main

import (
	"fmt"
)

func binarySearch(arr []int, target int) int {
	index := -1
	left := 0
	right := len(arr) - 1

	for left <= right {
		mid := (left + right) / 2

		if arr[mid] > target {
			right = mid
		}

		if arr[mid] < target {
			left = mid + 1
			index = mid
		}
	}

	return index
}

func main() {
	sample_array := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 15, 18, 21, 25, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50}
	target := 99

	fmt.Println("Sample Array:", sample_array)
	fmt.Println("Target:", target)

	index := binarySearch(sample_array, target)
	fmt.Println("Found index:", index, sample_array[index])
}
